File: spaces/shared/rendezvous.py
"""Which robots an account can reach, and whether each one is busy.

`GET /api/robot-status` with `Authorization: Bearer <hf token>`. Read off
`pollen-robotics/reachy_mini_central`'s `app.py` rather than guessed, because a `401` from it was
first mistaken here for something structural: the endpoint is `Depends(_resolve_hf_token)` and then
`validate_hf_token`, which is one `whoami-v2` call and **no scope check, no token-type check and no
requirement that the caller hold an event stream**. Its own docstring says what it is for — "render
a passive status indicator without consuming a session slot" — filtered to `p.username ==
username`, so an account sees its own robots and nobody else's.

Which makes it the right call for a listing, and better than the console's route for one:
`mediad/webclient/index.html` lists by opening `GET /events` and asking `list`, because a browser
that is about to start a session needs the stream anyway. Peers are keyed by token, so a second
`/events` on the same one supersedes the first (`remote-access-design.md` §3.7) — a page that
listed that way could not refresh its list without dropping the session it was holding. This one
opens nothing.

**A `429` here is not theirs.** `robot_status` is `validate_hf_token` and then a loop over the
producers — it never calls `check_rate_limit`, whose bucket is 1200 requests a minute keyed on a
hash of the token, and the only status it raises is `401`. So a 429 on this route was written by
something in front of the application, and `_who_answered` exists to say which something.

**A `401` here means the token, and nothing else.** Worth stating plainly because the first thing
that produced one was not a robot problem at all: run outside a Space, Gradio mocks its login and
hands the app the literal string `mock-oauth-token-for-local-dev`, which `whoami-v2` refuses
exactly as it should. `app.py`'s `token_of` is where that is dealt with.
"""

from __future__ import annotations

import logging
import os
from typing import Any

import requests

# The Space the mini's fleet registers with, and now ours: `mediad::relay::DEFAULT_RENDEZVOUS`.
#
# Written here rather than imported from `reachy_mini.media.central_consumer`, which is where it
# was read from originally. One constant is not worth a dependency on `aiortc`, `av` and a WebRTC
# stack in a Space that speaks only HTTP — and it made this module unimportable in the vision Space
# for no reason. `REACHY_CENTRAL_URL` overrides it, the same name their own `from_env` reads.
DEFAULT_CENTRAL_URL = os.environ.get(
    "REACHY_CENTRAL_URL", "https://pollen-robotics-reachy-mini-central.hf.space"
).rstrip("/")

# `meta.kind`, which is on the wire so one client can list two families of robot without opening a
# session to ask what it found. §5.1: their clients do not read it, this one does — an account with
# a duck and a mini on it would otherwise offer either as a duck.
DUCK = "microduck"

TIMEOUT = 20

# **What we call ourselves, and it is not cosmetic.** `requests` signs every call
# `python-requests/2.x`, which Hugging Face's edge treats as a bot: from a Space's container the
# very first `GET /api/robot-status` came back `429` with an HTML page and `server=awselb/2.0`,
# so the rendezvous never saw it. The reference client — `reachy_mini.media.central_consumer`,
# which `rf-detr-realtime-webcam` runs server-side from its own Space against this same host and
# route — sends no `User-Agent` of its own either, but it is built on `aiohttp` and inherits that
# library's signature instead. The difference between the two calls was the client, and nothing
# else: same URL, same `Authorization`, same everything.
#
# So this names the page honestly rather than imitating a browser. A caller that says who it is
# and carries a link is what a rate limiter is meant to let through, and it is the half of this
# that stays true if the rule ever changes.
USER_AGENT = (
    "microduck-policy-playground/1.0 "
    "(+https://huggingface.co/spaces/pollen-robotics/microduck-policy-playground)"
)

# Headers that say *who* answered. Asked only when the answer is one the application could not
# have written: an edge or a proxy names itself and carries an id worth quoting, where FastAPI
# would have sent `application/json` and a `detail`.
TELLTALE = (
    "retry-after",
    "server",
    "via",
    "x-request-id",
    "x-amzn-requestid",
    "cf-ray",
    "x-ratelimit-limit",
    "x-ratelimit-remaining",
    "x-ratelimit-reset",
    "content-type",
)


logger = logging.getLogger(__name__)


class RendezvousError(Exception):
    """Something the person who pressed the button can act on."""


class Robot:
    """One producer as the rendezvous describes it."""

    def __init__(self, entry: dict[str, Any]):
        meta = entry.get("meta") or {}
        self.peer_id: str = entry.get("peerId") or entry.get("id") or ""
        self.name: str = meta.get("name") or entry.get("robotName") or "a robot with no name"
        self.kind: str | None = meta.get("kind")
        self.release: str = meta.get("release") or "release unknown"
        # **A duck in MuJoCo, not one on a desk.** `mediad` sets this from `configd --simulated`;
        # `docs/design/simulation.md` §8 is why it is one declared fact rather than something each
        # client works out. Accepted as a bool or as the string a `GstStructure` turns it into,
        # because those are the two ways it reaches a listing and neither is this Space's choice.
        flag = meta.get("simulated")
        self.simulated: bool = flag is True or str(flag).lower() == "true"
        self.busy: bool = bool(entry.get("busy"))
        # Who has it, when it is busy. The rendezvous reports the *consumer's* name here, which
        # is why this Space sends a `consumer_label` worth reading.
        self.active_app: str | None = entry.get("activeApp")
        self.age: float | None = entry.get("last_seen_age_seconds")

    def label(self) -> str:
        """What the dropdown shows: enough to choose by, and the reason not to.

        `busy` matters more than it looks. One consumer at a time is the rendezvous's rule, so a
        robot somebody's console is already watching will refuse a session — and a dropdown that
        did not say so would make that look like a fault here.

        `simulated` matters for the opposite reason: nothing here goes wrong, and somebody drives a
        duck in MuJoCo believing it is the one on the shelf.
        """
        bits = [self.name]
        # Second, before anything about availability: which robot this is matters more than whether
        # it happens to be busy, and somebody scanning a dropdown of real ducks should not have to
        # read to the end of the line to find the one that is not real.
        if self.simulated:
            bits.append("simulated")
        bits.append(self.release)
        if self.busy:
            bits.append(f"busy with {self.active_app or 'something'}")
        if self.age is not None and self.age > 60:
            bits.append(f"last heard from {self.age / 60:.0f} min ago")
        return " — ".join(bits)


def _who_answered(answer: requests.Response) -> str:
    """Which hop produced this status, in the terms the answer itself offers.

    **A status the application cannot produce came from something in front of it**, and the only
    evidence of which something is what came back: a `server` that names itself, a request id to
    quote at whoever runs it, a `retry-after` that says whether this is a burst or a wall, and a
    body that is HTML where FastAPI would have written `{"detail": ...}`.

    The token is never part of this. What is quoted is the answer, which the caller already has.
    """
    bits = [f"{name}={answer.headers[name]}" for name in TELLTALE if name in answer.headers]
    body = " ".join((answer.text or "").split())[:200]
    if body:
        bits.append(f"body={body!r}")
    return ", ".join(bits) or "nothing — no telltale headers and an empty body"


def ducks(token: str, base: str = DEFAULT_CENTRAL_URL) -> tuple[list[Robot], list[str]]:
    """This account's ducks, and the names of whatever else was listed.

    Blocking: every caller is a Gradio callback with nothing else to do.
    """
    if not token:
        raise RendezvousError("no token to ask with")

    logger.info("GET %s/api/robot-status", base)
    try:
        answer = requests.get(
            f"{base}/api/robot-status",
            headers={"Authorization": f"Bearer {token}", "User-Agent": USER_AGENT},
            timeout=TIMEOUT,
        )
    except requests.RequestException as e:
        raise RendezvousError(f"the rendezvous could not be reached: {e}") from None

    logger.info("GET %s/api/robot-status -> HTTP %s", base, answer.status_code)
    # Logged for every refusal rather than only for the one that raises here, because which hop
    # answered is the question in all of them and the branches below each throw a different half
    # of it away.
    if answer.status_code != 200:
        logger.info("  answered by: %s", _who_answered(answer))
    if answer.status_code == 401:
        raise RendezvousError(
            "the rendezvous refused this token — `whoami-v2` did not recognise it. On a Space, "
            "sign in again. Running locally, Gradio's login button is a mock that hands the app "
            "a placeholder string, so the token comes from `HF_TOKEN` or `hf auth login` "
            "instead — the log line above says which one was used."
        )
    if answer.status_code == 429:
        raise RendezvousError(
            "something answered 429 before the rendezvous did — `/api/robot-status` raises 401 "
            "and nothing else, and its rate limiter is not on that route. So this is an edge in "
            f"front of the Space, and here is what it said: {_who_answered(answer)}"
        )
    if answer.status_code != 200:
        raise RendezvousError(
            f"the rendezvous answered HTTP {answer.status_code}: {answer.text[:200]}"
        )
    try:
        listed = answer.json().get("robots") or []
    except ValueError:
        raise RendezvousError("the rendezvous answered something that is not JSON") from None

    logger.info("%d producer(s) listed", len(listed))
    ours, theirs = [], []
    for entry in listed:
        robot = Robot(entry)
        logger.info(
            "  %s kind=%s busy=%s peer=%s", robot.name, robot.kind, robot.busy, robot.peer_id[:8]
        )
        if not robot.peer_id:
            continue
        if robot.kind == DUCK:
            ours.append(robot)
        else:
            theirs.append(f"{robot.name} ({robot.kind or 'no kind declared'})")
    return ours, theirs


if __name__ == "__main__":
    import os
    import sys

    from huggingface_hub import get_token

    logging.basicConfig(level=logging.INFO, format="%(name)s: %(message)s")
    credential = os.environ.get("HF_TOKEN") or get_token()
    if not credential:
        sys.exit("no token: run `hf auth login`, or set HF_TOKEN")
    try:
        found, others = ducks(credential)
    except RendezvousError as e:
        sys.exit(str(e))
    for duck in found:
        print(f"  {duck.peer_id}  {duck.label()}")
    print(f"\n{len(found)} duck(s)" + (f"; also listed: {', '.join(others)}" if others else ""))

File: spaces/policy-playground/web/src/rendezvous.ts
/**
 * JSON-RPC to a duck through the rendezvous, from the browser.
 *
 * A direct port of `spaces/shared/wire.py`, which is the transport that works from anywhere: no
 * WebRTC, no ICE, no relay candidate. The service forwards every key of a `peer` envelope except
 * `type` and `sessionId` verbatim to the session partner, so an envelope carrying `rpc` is a
 * control call and `mediad::relay`'s control lane answers it out of the routing table the
 * datachannel uses.
 *
 *     POST /send  {"type": "peer", "sessionId": S, "rpc": {"jsonrpc": "2.0", "id": 1, …}}
 *     SSE         {"type": "peer", "sessionId": S, "rpc": {"jsonrpc": "2.0", "id": 1, "result": …}}
 *
 * Four things this protocol punishes a reader for not knowing, all learned the hard way in the
 * Python version and all still true here:
 *
 * - **`POST /send` before `GET /events` is a 400.** Identity comes from the bearer token and the
 *   stream is what binds it, so the stream opens first and everything waits for the `welcome`.
 * - **`startSession` and `list` answer in the POST body**, not on the stream. Everything else
 *   arrives on the stream. This is the shape a reader gets wrong once.
 * - **CRLF is not ours to assume.** SSE permits `\r\n` and a proxy may rewrite them; splitting on
 *   `\n\n` alone then matches nothing and every message vanishes in silence.
 * - **One consumer per robot.** `sessionRejected` means somebody else holds it, and the robot's
 *   own console counts.
 *
 * And one thing the browser changes for the better: `EventSource` cannot send an `Authorization`
 * header, so this reads the stream with `fetch` and a reader. That is not a workaround — it is
 * what lets the token stay in a header rather than a query string the service only kept as a
 * deprecated fallback.
 */

export const DEFAULT_CENTRAL_URL = "https://pollen-robotics-reachy-mini-central.hf.space";

/** `meta.kind`, which is how one client tells a duck from a mini without opening a session. */
const DUCK = "microduck";

export class RendezvousError extends Error {}

export interface Robot {
  peerId: string;
  name: string;
  kind: string | null;
  release: string;
  busy: boolean;
  activeApp: string | null;
  ageSeconds: number | null;
}

interface Envelope {
  type?: string;
  [key: string]: unknown;
}

function robotFrom(entry: Record<string, unknown>): Robot {
  const meta = (entry.meta ?? {}) as Record<string, unknown>;
  return {
    peerId: String(entry.peerId ?? entry.id ?? ""),
    name: String(meta.name ?? entry.robotName ?? "a robot with no name"),
    kind: meta.kind == null ? null : String(meta.kind),
    release: String(meta.release ?? "release unknown"),
    busy: Boolean(entry.busy),
    activeApp: entry.activeApp == null ? null : String(entry.activeApp),
    ageSeconds: typeof entry.last_seen_age_seconds === "number" ? entry.last_seen_age_seconds : null,
  };
}

/**
 * The ducks this account can reach, and whatever else was listed.
 *
 * `GET /api/robot-status` is one request and opens no event stream, so it is safe to press
 * mid-session — unlike the console's `list`, which would open a second stream on the same token
 * and evict the peer the session is riding on.
 */
export async function listDucks(token: string, base = DEFAULT_CENTRAL_URL): Promise<{
  ducks: Robot[];
  others: string[];
}> {
  if (!token) throw new RendezvousError("no token to ask with");

  let answer: Response;
  try {
    answer = await fetch(`${base}/api/robot-status`, {
      headers: { Authorization: `Bearer ${token}` },
    });
  } catch (e) {
    throw new RendezvousError(`the rendezvous could not be reached: ${e}`);
  }

  if (answer.status === 401) {
    throw new RendezvousError(
      "the rendezvous did not recognise this sign-in. Signing out and back in is the fix.",
    );
  }
  if (!answer.ok) {
    throw new RendezvousError(
      `the rendezvous answered HTTP ${answer.status}: ${(await answer.text()).slice(0, 200)}`,
    );
  }

  const body = (await answer.json()) as { robots?: unknown };
  const listed = Array.isArray(body.robots) ? body.robots : [];
  const ducks: Robot[] = [];
  const others: string[] = [];
  for (const entry of listed) {
    if (typeof entry !== "object" || entry === null) continue;
    const robot = robotFrom(entry as Record<string, unknown>);
    if (!robot.peerId) continue;
    if (robot.kind === DUCK) ducks.push(robot);
    else others.push(`${robot.name} (${robot.kind ?? "no kind declared"})`);
  }
  return { ducks, others };
}

type Pending = {
  method: string;
  resolve: (value: unknown) => void;
  reject: (reason: Error) => void;
  timer: number;
};

export class RpcError extends Error {
  constructor(
    readonly method: string,
    message: string,
  ) {
    super(message);
  }
}

/** How long a call may go unanswered. `policy.fetch` overrides it — that one is a download. */
const CALL_TIMEOUT = 20_000;

/**
 * One control-only session with one duck.
 *
 * `onEvent` is how the page learns what is happening without polling: every envelope, every call
 * and every refusal is announced, which is what the log panel reads and what the status chip
 * reads. A transport that only reported success would leave four different silences looking
 * identical, which is the failure this whole file was written twice to avoid.
 */
export class Session {
  private stream: AbortController | null = null;
  private sessionId: string | null = null;
  private selfPeerId: string | null = null;
  private nextId = 1;
  private readonly pending = new Map<number, Pending>();
  private welcome: ((message: Envelope) => void) | null = null;

  constructor(
    private readonly token: string,
    private readonly peerId: string,
    private readonly label: string,
    private readonly onEvent: (line: string) => void,
    private readonly base = DEFAULT_CENTRAL_URL,
  ) {}

  get open(): boolean {
    return this.sessionId !== null;
  }

  async start(): Promise<void> {
    const welcomed = new Promise<Envelope>((resolve, reject) => {
      this.welcome = resolve;
      window.setTimeout(
        () => reject(new RendezvousError("the rendezvous accepted the stream and never said hello")),
        20_000,
      );
    });

    this.openStream();
    const hello = await welcomed;
    this.selfPeerId = typeof hello.peerId === "string" ? hello.peerId : null;
    this.onEvent(`welcome: peer ${(this.selfPeerId ?? "?").slice(0, 8)}`);

    // A name in the listing, so the owner's other devices can see who holds the robot — the
    // service reports a consumer's `meta.name` back as `activeApp`.
    await this.post({ type: "setPeerStatus", roles: ["listener"], meta: { name: this.label } });

    // The session is what makes a `peer` envelope routable: the service drops one naming a session
    // it does not know. Nothing about it commits either end to WebRTC.
    const answer = await this.post({ type: "startSession", peerId: this.peerId });
    const kind = answer?.type;
    if (kind === "sessionRejected") {
      throw new RendezvousError(
        `this duck is busy with ${String(answer?.activeApp ?? "something else")} — ` +
          "one at a time, and the robot's own console counts.",
      );
    }
    if (kind === "error") {
      throw new RendezvousError(String(answer?.details ?? "the rendezvous refused"));
    }
    if (kind !== "sessionStarted") {
      throw new RendezvousError(`startSession answered ${JSON.stringify(answer)}`);
    }
    this.sessionId = String(answer?.sessionId ?? "");
    this.onEvent(`session ${this.sessionId.slice(0, 8)} open, control only`);
  }

  async stop(): Promise<void> {
    const id = this.sessionId;
    this.sessionId = null;
    if (id) {
      // A teardown that cannot be delivered still ends this side: the service drops the robot's
      // lane on its own endSession broadcast either way.
      await this.post({ type: "endSession", sessionId: id }).catch(() => undefined);
    }
    this.stream?.abort();
    this.stream = null;
    this.abandon("disconnected");
  }

  /** One JSON-RPC request, answered on the stream. */
  async call(method: string, params: Record<string, unknown> = {}, timeout = CALL_TIMEOUT): Promise<unknown> {
    if (!this.sessionId) throw new RpcError(method, "not connected to a duck");
    const id = this.nextId++;
    this.onEvent(`→ ${method} ${JSON.stringify(params)}`.slice(0, 300));

    const answered = new Promise<unknown>((resolve, reject) => {
      const timer = window.setTimeout(() => {
        this.pending.delete(id);
        reject(new RpcError(method, `no answer in ${Math.round(timeout / 1000)}s`));
      }, timeout);
      this.pending.set(id, { method, resolve, reject, timer });
    });

    await this.post({
      type: "peer",
      sessionId: this.sessionId,
      rpc: { jsonrpc: "2.0", id, method, params },
    });
    return answered;
  }

  private openStream(): void {
    const controller = new AbortController();
    this.stream = controller;
    void (async () => {
      try {
        const answer = await fetch(`${this.base}/events`, {
          headers: { Authorization: `Bearer ${this.token}`, Accept: "text/event-stream" },
          signal: controller.signal,
        });
        if (answer.status === 401) throw new RendezvousError("the rendezvous refused this sign-in");
        if (!answer.ok || !answer.body) throw new RendezvousError(`the event stream answered HTTP ${answer.status}`);

        const reader = answer.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";
        for (;;) {
          const { done, value } = await reader.read();
          if (done) break;
          // CRLF normalised on arrival: the framing is not ours, and a proxy may rewrite it.
          buffer += decoder.decode(value, { stream: true }).replaceAll("\r\n", "\n");
          let cut = buffer.indexOf("\n\n");
          while (cut !== -1) {
            const frame = buffer.slice(0, cut);
            buffer = buffer.slice(cut + 2);
            const data = frame
              .split("\n")
              .filter((line) => line.startsWith("data:"))
              .map((line) => line.slice(5).trim())
              .join("");
            if (data) this.handle(data);
            cut = buffer.indexOf("\n\n");
          }
        }
        if (!controller.signal.aborted) this.ended("the rendezvous closed the event stream");
      } catch (e) {
        // Aborting mid-read is how `stop` works, and it throws. That is a clean teardown and not
        // something to report as a failure.
        if (!controller.signal.aborted) this.ended(`the event stream failed: ${e}`);
      }
    })();
  }

  private handle(data: string): void {
    let message: Envelope;
    try {
      message = JSON.parse(data) as Envelope;
    } catch {
      this.onEvent(`← unparseable frame: ${data.slice(0, 160)}`);
      return;
    }

    if (message.type === "welcome") {
      this.welcome?.(message);
      this.welcome = null;
      return;
    }

    if (message.type === "peer") {
      const payload = message.rpc as Record<string, unknown> | undefined;
      // An `sdp` or an `ice`: a media negotiation this transport never started. Not an error.
      if (!payload) return;
      const id = typeof payload.id === "number" ? payload.id : null;
      if (id === null) return;
      const waiting = this.pending.get(id);
      if (!waiting) return;
      this.pending.delete(id);
      window.clearTimeout(waiting.timer);
      const failure = payload.error as { message?: unknown } | undefined;
      if (failure) {
        this.onEvent(`← ${waiting.method} refused: ${String(failure.message ?? "no reason")}`);
        waiting.reject(new RpcError(waiting.method, String(failure.message ?? "refused")));
      } else {
        this.onEvent(`← ${waiting.method} ${JSON.stringify(payload.result).slice(0, 200)}`);
        waiting.resolve(payload.result);
      }
      return;
    }

    if (message.type === "endSession" || message.type === "sessionRejected") {
      this.ended(`the session ended: ${String(message.reason ?? "no reason given")}`);
    }
  }

  private ended(why: string): void {
    this.sessionId = null;
    this.onEvent(why);
    this.abandon(why);
  }

  private abandon(why: string): void {
    for (const [id, waiting] of this.pending) {
      window.clearTimeout(waiting.timer);
      waiting.reject(new RpcError(waiting.method, why));
      this.pending.delete(id);
    }
  }

  private async post(message: Envelope): Promise<Envelope | null> {
    let answer: Response;
    try {
      answer = await fetch(`${this.base}/send`, {
        method: "POST",
        headers: { Authorization: `Bearer ${this.token}`, "Content-Type": "application/json" },
        body: JSON.stringify(message),
      });
    } catch (e) {
      throw new RendezvousError(`POST /send: ${e}`);
    }
    if (answer.status === 429) {
      throw new RendezvousError(
        "the rendezvous is rate-limiting this sign-in (1200 requests a minute).",
      );
    }
    if (answer.status === 400) {
      throw new RendezvousError(
        "the rendezvous says this peer does not exist — its event stream is gone. Connect again.",
      );
    }
    if (!answer.ok) {
      throw new RendezvousError(`POST /send answered HTTP ${answer.status}`);
    }
    try {
      const body = (await answer.json()) as Envelope;
      // `startSession` and `list` answer here; everything else answers on the stream.
      return body && body.type ? body : null;
    } catch {
      return null;
    }
  }
}
